core/libp2p/protocol_muxer/multiselect/multiselect.cpp

```diff
@@ -5,269 +5,257 @@
-                                        response.protocols[0],
-                                        previous_protocol, current_status));
-          current_status = status;
-          if (current_status == Status::NEGOTIATION_SUCCESS) {
-            previous_protocol = response.protocols[0];
-          }
-          break;
-        }
-        case MessageType::PROTOCOLS: {
-          OUTCOME_TRY(status,
-                      handleProtocolsMsg(connection, supported_protocols,
-                                         response.protocols, current_status));
-          current_status = status.first;
-          if (current_status == Status::PROTOCOL_SENT) {
-            previous_protocol = status.second;
-          }
-          break;
-        }
-        case MessageType::LS: {
-          OUTCOME_TRY(status, handleLsMsg(connection, supported_protocols));
-          current_status = status;
-          if (current_status == Status::PROTOCOL_SENT) {
-            // memorize the protocol we sent
-            previous_protocol = response.protocols[0];
-          }
-          break;
-        }
-        case MessageType::NA: {
-          OUTCOME_TRY(status, handleNaMsg(connection));
-          current_status = status;
-          break;
-        }
-        default:
-          return MultiselectError::INTERNAL_ERROR;
-      }
-    }
-    return finalizeNegotiation(current_status, previous_protocol);
+  void Multiselect::onWriteAckCompleted(
+      const std::shared_ptr<ConnectionState> &connection_state,
+      const Protocol &protocol) {
+    negotiationRoundFinished(connection_state, protocol);
   }
 
-  outcome::result<peer::Protocol> Multiselect::finalizeNegotiation(
-      const Status status, const peer::Protocol &protocol) const {
-    switch (status) {
-      case Status::NEGOTIATION_SUCCESS:
-        return protocol;
-      case Status::NEGOTIATION_FAIL:
-        return MultiselectError::NEGOTIATION_FAILED;
+  void Multiselect::onReadCompleted(
+      std::shared_ptr<ConnectionState> connection_state,
+      MessageManager::MultiselectMessage msg) {
+    using MessageType = MessageManager::MultiselectMessage::MessageType;
+
+    switch (msg.type) {
+      case MessageType::OPENING:
+        handleOpeningMsg(std::move(connection_state));
+        return;
+      case MessageType::PROTOCOL:
+        handleProtocolMsg(msg.protocols[0], connection_state);
+        return;
+      case MessageType::PROTOCOLS:
+        handleProtocolsMsg(msg.protocols, connection_state);
+        return;
+      case MessageType::LS:
+        handleLsMsg(connection_state);
+        return;
+      case MessageType::NA:
+        handleNaMsg(connection_state);
+        return;
       default:
-        // never should be here
-        log_->critical("achieved state, which never should be achievable");
-        return MultiselectError::INTERNAL_ERROR;
+        log_->critical(
+            "type of the message, returned by the parser, is unknown");
+        return;
```

Comment:
You must not just return from any point of Multiselect - `clearResources` and call a `proto_callback`, as you do in, for example, `onGarbagedStreamStatus`. Otherwise, the client of Multiselect will wait forever for negotiation result